homeassistant/components/shelly/update.py

```diff
@@ -0,0 +1,236 @@
+            "version"
+        ],
+        install=lambda wrapper: wrapper.async_trigger_ota_update(),
+        device_class=UpdateDeviceClass.FIRMWARE,
+        entity_category=EntityCategory.CONFIG,
+        entity_registry_enabled_default=True,
+    ),
+    "fwupdate_beta": RpcUpdateDescription(
+        name="Beta Firmware Update",
+        key="sys",
+        sub_key="available_updates",
+        latest_version=lambda status: status.get("beta", {"version": None})["version"],
+        install=lambda wrapper: wrapper.async_trigger_ota_update(beta=True),
+        device_class=UpdateDeviceClass.FIRMWARE,
+        entity_category=EntityCategory.CONFIG,
+        entity_registry_enabled_default=False,
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up update entities for Shelly component."""
+    if get_device_entry_gen(config_entry) == 2:
+        return async_setup_entry_rpc(
+            hass, config_entry, async_add_entities, RPC_UPDATES, RpcUpdateEntity
+        )
+
+    if not config_entry.data[CONF_SLEEP_PERIOD]:
+        async_setup_entry_rest(
+            hass,
+            config_entry,
+            async_add_entities,
+            REST_UPDATES,
+            RestUpdateEntity,
+        )
+
+
+class RestUpdateEntity(ShellyRestAttributeEntity, UpdateEntity):
+    """Represent a REST update entity."""
+
+    _attr_supported_features = (
+        UpdateEntityFeature.INSTALL | UpdateEntityFeature.PROGRESS
+    )
+    entity_description: RestUpdateDescription
+
+    def __init__(
+        self,
+        wrapper: BlockDeviceWrapper,
+        attribute: str,
+        description: RestEntityDescription,
+    ) -> None:
+        """Initialize update entity."""
+        super().__init__(wrapper, attribute, description)
+        self._in_progress_old_version: str | None = None
+
+    @property
+    def installed_version(self) -> str | None:
+        """Version currently in use."""
+        version = self.wrapper.device.status["update"]["old_version"]
+        if version is None:
+            return None
+
+        return cast(str, version)
+
+    @property
+    def latest_version(self) -> str | None:
+        """Latest version available for install."""
+        new_version = self.entity_description.latest_version(
+            self.wrapper.device.status,
+        )
+        if new_version is not None:
+            return cast(str, new_version)
+
+        return self.installed_version
+
+    @property
+    def in_progress(self) -> bool:
+        """Update installation in progress."""
+        return self._in_progress_old_version == self.installed_version
+
+    async def async_install(
+        self, version: str | None, backup: bool, **kwargs: Any
+    ) -> None:
+        """Install the latest firmware version."""
+        config_entry = self.wrapper.entry
+        block_wrapper = self.hass.data[DOMAIN][DATA_CONFIG_ENTRY][
```

Comment:
Is this a different wrapper than `self.wrapper`?

Reply:
Yes, the self.wrapper is the REST wrapper and for issuing the update request we need the BLOCK wrapper.

Reply:
Ok. Consider updating the design to pass the block wrapper to the entity when creating it. Accessing hass.data here doesn't seem like the best design.

Reply:
I will look into it.

Reply:
@MartinHjelmare I implemented this in PR: https://github.com/home-assistant/core/pull/79297
Could you review that?